Don't use intermediate list in decodeContainerSkel

michalt opened the pull request:
As I've mentioned in the commit description, this doesn't seem to change benchmark results. But it still seems that the code is a bit more readable.

arianvp commented:
This seems similar in goal as https://github.com/well-typed/binary-serialise-cbor/pull/36  .  Shall we close https://github.com/well-typed/binary-serialise-cbor/pull/36 in favor of this pull?

michalt commented:
Oh, I completely missed your PR! (I only pulled master to make sure that I haven't missed anything...) Sorry for that!

Anyway, I'm also using `replicateM` for `Sequence` and refactoring things a bit to make them a bit more readable. If you think that supersedes your PR, feel free to close it. Otherwise, let's wait to see what the maintainers have in mind. :-)

dcoutts commented:
I'm somewhat concerned about using this approach directly because it's vulnerable to DOS attacks when using untrusted input. The attacker simply needs to claim the sequence is a bazillion long and it'll go and allocate a vector of that size up front.

Can we make ourselves a variant that is able to go in chunks? What might be useful here is a way in the decoder to ask "how much data is pending?" so we can make sure we don't allocate chunks that are much bigger than the amount of data the user (or attacker) has supplied.

arianvp commented:
@dcoutts  something like this (Typed on my mobile phone, so maybe some typo's) ? 

It will  build up `chunkSize` sized vectors and combines them afterwards.  

    chunkedReplicateM :: Int -> Int -> m a -> m (Vector a)
    chunkedReplicateM chunkSize size = 
        let (n, r) = divMod size chunkSize
        in  concatMap (\v -> replicateN r m) (fmap join (replicateM n (replicateM chunkSize m)))

michalt commented:
@dcoutts @arianvp Added the chunking based on your ideas - create a list of containers,  add a `DecodeAction` to have a look at the length of the current `ByteString`. Let me know what you think about this!

arianvp commented on Data/Binary/Serialise/CBOR/Class.hs:
```diff
@@ -269,11 +269,30 @@ encodeContainerSkel encodeLen size foldr f  c =
+       then replicateFun size decode
+       else do
+           -- Use at least 128 as the size of the chunk.
+           let chunkSize = max limit 128
```
So the chunk size grows as the amount of data that we decode grows?

dcoutts commented:
Yes, I think an approach like this makes sense. The peek length primitive is sensible, though we should clarify if it's bytes or words or what. Perhaps it should itself return a minimum, since it'll sometimes be unlucky and be called just near the end of a chunk.

BTW, I suspect the reason that the benchmarks don't improve is because Vector.replicateM in fact does allocate lists whenever the monad it's executed in is not IO or ST. I may be mis-remembering but I did chase definitions within vector once and found that it would indeed fall back to using lists. May be worth checking in this case. We may have to either provide the ability to use ST or to build array construction into the decoder interpreter.

michalt commented:
I've added some comments to clarify the bytes vs elements. I'm not sure I understand what you mean by:

> Perhaps it should itself return a minimum, since it'll sometimes be unlucky and be called just near the end of a chunk.

In particular what should return the minimum (or where should we use it)? I've picked using maximum of the "peeked bytes" and a constant because there can be few bytes remaining (if we're at the end of the chunk). In any way, it's just a heuristic and I'll be happy to change this if someone has better ideas.

I've also found and fixed a bug with `PeekLength` which for `go_fast_end` should be called as one of the actions that don't need any input (fixed by moving this case earlier in pattern matching).

Re performance: now it seems that there is a reproducible ~5% improvement. Maybe connected with #42 ? (the commit to `INLINE` the `encode` and `decode` for `Vector` instances is already in)

michalt commented:
Oh, and I'll try to have a look next week at `vector` to see if it's using lists for monads other than `IO` or `ST`.

arianvp commented:
Would running the benchmark in the IO monad / ST Monad mitigate this?

arianvp commented:
What still needs to happen to get this merged?

dcoutts commented:
@thoughtpolice do you have any time to think about this? I'm kind of swamped at the moment.

thoughtpolice commented:
Hey, I'm on a tear to get rid of all the PRs, including this one! I took a look a few days ago but still haven't come back to it. I think it might be tentatively OK - @michalt if you get around to rebasing this before I do that would be useful, since the CI will now give you a pass!

michalt commented:
Sounds good, will rebase soon.

michalt commented:
@thoughtpolice should be ready :-)

thoughtpolice commented:
Thanks! I split your commits up slightly differently, but it's all there. Sorry for the wait on this one!

michalt commented:
No worries, thanks for merging :-)
